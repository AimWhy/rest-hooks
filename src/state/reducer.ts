import { normalize } from '../resource';
import { merge } from 'lodash';
import { Resource } from '../resource';
import { ActionTypes, State } from '../types';

export const initialState: State<Resource> = {
  entities: {},
  results: {},
  meta: {},
};

export default function reducer(state: State<Resource>, action: ActionTypes) {
  switch (action.type) {
    case 'receive':
      if (action.error) {
        return {
          ...state,
          meta: {
            ...state.meta,
            [action.meta.url]: {
              date: action.meta.date,
              error: action.payload,
              expiresAt: action.meta.expiresAt,
            },
          },
        };
      }
      const normalized = normalize(action.payload, action.meta.schema);
      return {
        entities: merge({ ...state.entities }, normalized.entities),
        results: {
          ...state.results,
          [action.meta.url]: normalized.result,
        },
        meta: {
          ...state.meta,
          [action.meta.url]: {
            date: action.meta.date,
            expiresAt: action.meta.expiresAt,
          },
        },
      };
    case 'rpc':
      const { entities } = normalize(action.payload, action.meta.schema);
      return {
        ...state,
        entities: merge({ ...state.entities }, entities),
      };
    default:
      // A reducer must always return a valid state.
      // Alternatively you can throw an error if an invalid action is dispatched.
      return state;
  }
}
