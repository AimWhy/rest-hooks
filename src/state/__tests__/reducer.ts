import { ArticleResource } from '../../__tests__/common';
import reducer from '../reducer';

describe('reducer', () => {
  it('handle get single', () => {
    const id = 20;
    const action = {
      type: 'receive' as 'receive',
      payload: { id, title: 'hi', content: 'this is the content' },
      meta: {
        schema: ArticleResource.getSchema(),
        url: ArticleResource.url({ id }),
        mutate: false,
        date: 5000000000,
        expiresAt: 5000500000,
      },
    };
    const iniState = {
      entities: {},
      results: {},
      meta: {},
    };
    const newState = reducer(iniState, action);
    expect(newState).toMatchSnapshot();
  });
  it('mutate should never change results', () => {
    const id = 20;
    const action = {
      type: 'receive' as 'receive',
      payload: { id, title: 'hi', content: 'this is the content' },
      meta: {
        schema: ArticleResource.getSchema(),
        url: ArticleResource.url({ id }),
        mutate: true,
        date: 5000000000,
        expiresAt: 5000500000,
      },
    };
    const iniState = {
      entities: {},
      results: { abc: '5' },
      meta: {},
    };
    const newState = reducer(iniState, action);
    expect(newState.results).toBe(iniState.results);
  });
  it('should set error in meta', () => {
    const id = 20;
    const error = new Error('hi');
    const action = {
      type: 'receive' as 'receive',
      payload: error,
      meta: {
        schema: ArticleResource.getSchema(),
        url: ArticleResource.url({ id }),
        mutate: false,
        date: 5000000000,
        expiresAt: 5000500000,
      },
      error: true,
    };
    const iniState = {
      entities: {},
      results: {},
      meta: {},
    };
    const newState = reducer(iniState, action);
    expect(newState).toMatchSnapshot();
  });
  it('other types should do nothing', () => {
    const action = {
      type: 'fetch' as 'fetch',
      payload: () => new Promise<any>(() => null),
      meta: {
        schema: ArticleResource.getSchema(),
        url: ArticleResource.url({ id: 5 }),
        mutate: true,
        throttle: true,
        reject: (v: any) => null,
        resolve: (v: any) => null,
      },
    };
    const iniState = {
      entities: {},
      results: { abc: '5' },
      meta: {},
    };
    const newState = reducer(iniState, action);
    expect(newState).toBe(iniState);
  })
});
