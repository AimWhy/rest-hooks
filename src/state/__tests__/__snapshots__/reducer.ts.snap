// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer handle get single 1`] = `
Object {
  "entities": Object {
    "http://test.com/article/": Object {
      "20": ArticleResource {
        "author": null,
        "content": "this is the content",
        "id": 20,
        "tags": Array [],
        "title": "hi",
      },
    },
  },
  "meta": Object {
    "http://test.com/article/20": Object {
      "date": 5000000000,
      "expiresAt": 5000500000,
    },
  },
  "results": Object {
    "http://test.com/article/20": "20",
  },
}
`;

exports[`reducer should set error in meta 1`] = `
Object {
  "entities": Object {},
  "meta": Object {
    "http://test.com/article/20": Object {
      "date": 5000000000,
      "error": [Error: hi],
      "expiresAt": 5000500000,
    },
  },
  "results": Object {},
}
`;
